### PR TITLE
Course Discovery (includes a docker devstack)

### PR DESCRIPTION
**Pull request by cpennington**
This builds on hackathon work to enable docker to build a new IDA that uses docker for its devstack.

This should be rebased after #2499, #2500, and #2501 are reviewed and merged.

### CODE REVIEW

**Comment by cpennington** [comment]
@e0d: Review?

**Comment by cpennington** [comment]
Also, it seems like squashing many of these commits would be appropriate.

File: .dockerignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+*
+!playbooks/
+!docker/
```
Comment (clintonb) [inline .dockerignore]: Add a newline.

File: docker/build/course-discovery/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# To build this Dockerfile:
+#
+# From the root of configuration:
+#
+# docker build -f docker/build/course-discovery/Dockerfile .
+#
+# This allows the dockerfile to updated /edx/app/edx_ansible/edx_ansible
```
Comment (clintonb) [inline docker/build/course-discovery/Dockerfile]: updated -> update

File: .dockerignore (hunk 1 above)
Comment (cpennington) [inline .dockerignore]: Done.

File: docker/build/course-discovery/Dockerfile (hunk 2 above)
Comment (cpennington) [inline docker/build/course-discovery/Dockerfile]: Done

**Comment by e0d** [comment]
@cpennington planning to review tomorrow in the hack.  Noticed that there's a syntax issue your last commit, see the failed check.

File: docker/build/mysql/edxlocal.yml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# Configure a Jenkins worker instance
```
Comment (e0d) [inline docker/build/mysql/edxlocal.yml]: where's this comment block from?

File: docker/plays/course_discovery.yml [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  hosts: all
+  sudo: True
+  gather_facts: True
+  vars:
+    serial_count: 1
+  serial: "{{ serial_count }}"
+  roles:
+    - docker
+    - role: course_discovery
+      nginx_default_sites:
```
Comment (e0d) [inline docker/plays/course_discovery.yml]: Nginx is running on the CD container?

File: playbooks/filter_plugins/ports_plugin.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import hashlib
```
Comment (e0d) [inline playbooks/filter_plugins/ports_plugin.py]: Can you provide a comment for the rationale for pulling this in and what needs to happen to remove it..

Oh wait, this is the proposed hash based generation of ports.  I'm not sure I favour this.  I think a more simplistic /etc/services approach makes more sense.

File: playbooks/roles/container/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,26 @@
+---
+#
```
Comment (e0d) [inline playbooks/roles/container/tasks/main.yml]: we should delete before merging.

File: playbooks/roles/course_discovery/defaults/main.yml [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+COURSE_DISCOVERY_CACHES:
+  default:
+    BACKEND:  'django.core.cache.backends.memcached.MemcachedCache'
+    KEY_PREFIX: 'default'
+    LOCATION: memcache
+
+COURSE_DISCOVERY_VERSION: "master"
+COURSE_DISCOVERY_DJANGO_SETTINGS_MODULE: "edx_course_discovery.settings.production"
+COURSE_DISCOVERY_URL_ROOT: 'http://localhost:8381'
```
Comment (e0d) [inline playbooks/roles/course_discovery/defaults/main.yml]: I'd like to remove all localhost references to in favour of something like coursediscover.edx.

File: playbooks/roles/course_discovery/defaults/main.yml [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+course_discovery_code_dir: "{{ course_discovery_home }}/{{ course_discovery_service_name }}"
+
+course_discovery_gunicorn_host: "127.0.0.1"
+course_discovery_gunicorn_port: 8381
+course_discovery_gunicorn_timeout: 300
+
+course_discovery_log_dir: "{{ COMMON_LOG_DIR }}/{{ course_discovery_service_name }}"
+
+course_discovery_requirements_base: "{{ course_discovery_code_dir }}/requirements"
+course_discovery_requirements:
```
Comment (e0d) [inline playbooks/roles/course_discovery/defaults/main.yml]: Would love to remove this in favour of `make requrements`

File: playbooks/roles/course_discovery/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+- name: add gunicorn configuration file
+  template:
+    src: edx/app/course_discovery/course_discovery_gunicorn.py.j2
+    dest: "{{ course_discovery_home }}/course_discovery_gunicorn.py"
+  sudo_user: "{{ course_discovery_user }}"
+  tags:
+    - install
+    - install:configuration
+
+- name: add deadsnakes repository
```
Comment (e0d) [inline playbooks/roles/course_discovery/tasks/main.yml:33]: @feanil thoughts on whether we should factor this out into  a role, there's not much to it, but we're likely to reuse a lot.

File: playbooks/roles/course_discovery/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+  apt_repository:
+    repo: "ppa:fkrull/deadsnakes"
+  tags:
+    - install
+    - install:system-requirements
+
+- name: install python3.5
```
Comment (e0d) [inline playbooks/roles/course_discovery/tasks/main.yml:40]: why hardcode 3.5

File: docker/build/mysql/edxlocal.yml (hunk 3 above)
Comment (cpennington) [inline docker/build/mysql/edxlocal.yml]: This was part of the pre-existing docker work. My container doesn't depend on it.

File: docker/plays/course_discovery.yml (hunk 4 above)
Comment (cpennington) [inline docker/plays/course_discovery.yml]: My goal was to have the docker devstack running production-style, as far as access to static assets and the like go, so, yes, I put nginx in the same container.

File: playbooks/filter_plugins/ports_plugin.py (hunk 5 above)
Comment (cpennington) [inline playbooks/filter_plugins/ports_plugin.py]: The goal here is to have fewer steps from running the IDA generation script to actually having a deployment up and running. Rather than having to search through and specify a bunch of ports, this simply creates them for you.

That said, I'd be ok with having the IDA creation script just ask for a port suffix as an argument.

File: playbooks/roles/container/tasks/main.yml (hunk 6 above)
Comment (cpennington) [inline playbooks/roles/container/tasks/main.yml]: Happily. No idea where that came from (I assume pre-existing docker work).

File: playbooks/roles/course_discovery/defaults/main.yml (hunk 7 above)
Comment (cpennington) [inline playbooks/roles/course_discovery/defaults/main.yml]: Ah, an excellent point.

File: playbooks/roles/course_discovery/defaults/main.yml (hunk 8 above)
Comment (cpennington) [inline playbooks/roles/course_discovery/defaults/main.yml]: Works for me.

File: playbooks/roles/course_discovery/tasks/main.yml (hunk 9 above)
Comment (cpennington) [inline playbooks/roles/course_discovery/tasks/main.yml:33]: My thinking was just that for now, the python3 stuff is a bit experimental (gevent and gunicorn don't yet have released/stable python 3 support). So, until it's more stable, I'd rather have it specific to course-discovery (that's also why it's not in the template code).

File: playbooks/roles/course_discovery/tasks/main.yml (hunk 10 above)
Comment (cpennington) [inline playbooks/roles/course_discovery/tasks/main.yml:40]: See my above note. This is experimental.

**Comment by cpennington** [comment]
@e0d: I don't think the error is a syntax error that's breaking the build, it's that the `filter_plugins` directory isn't being found when jenkins runs. But I'm not sure why that is.

File: docker/plays/course_discovery.yml (hunk 4 above)
Comment (e0d) [inline docker/plays/course_discovery.yml]: Is there something more complex than proxying to the backend going on here?

Comment (cpennington) [inline docker/plays/course_discovery.yml]: No, but that's partly since the app doesn't have any assets yet. I expect once we have assets, we'll want nginx to serve them.

File: playbooks/roles/course_discovery/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+  sudo_user: "{{ course_discovery_user }}"
+  with_items: "{{ course_discovery_requirements }}"
+  tags:
+    - install
+    - install:app-requirements
+
+- name: wait for database
+  wait_for: host=db port={{ COURSE_DISCOVERY_DATABASES.default.PORT }} delay=2
+  tags:
+    - migrate:devstack
```
Comment (e0d) [inline playbooks/roles/course_discovery/tasks/main.yml]: maybe this should just be install:db.  There's not reason this could not be used to bootstrap any deployment of this IDA is there?

File: vagrant/base/devstack/Vagrantfile [hunk 12]
```diff
@@ -8,32 +8,25 @@ VAGRANTFILE_API_VERSION = "2"
 MEMORY = 4096
```
Comment (e0d) [inline vagrant/base/devstack/Vagrantfile]: Can you pull these changes into a separate PR?  We're going to need to tease this changeset apart generally.  I'd like to pull in the related pieces as discrete PRs and then have you rebase your PR on the new master.

File: playbooks/roles/course_discovery/tasks/main.yml (hunk 11 above)
Comment (cpennington) [inline playbooks/roles/course_discovery/tasks/main.yml]: Yeah, that would be fine too. I just wasn't sure whether we wanted this type of bootstrapping (and permissions) elsewhere, or preferred a different bootstrapping style.

File: vagrant/base/devstack/Vagrantfile (hunk 12 above)
Comment (cpennington) [inline vagrant/base/devstack/Vagrantfile]: Yes, I can do that.

File: docker/build/course_discovery/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+ARG course_discovery_version=master
+ARG repo_owner=edx
+
+ADD . /edx/app/edx_ansible/edx_ansible
+
+USER docker
+WORKDIR /edx/app/edx_ansible/edx_ansible/docker/plays
+
+RUN sudo ansible-playbook course_discovery.yml -c local -t 'install:base'
```
Comment (feanil) [inline docker/build/course_discovery/Dockerfile]: Why are you running all of these tags separately?  Is there a tag you're avoiding?  Also why not as a single command?

File: docker/compose/docker-compose.yml [hunk 14]
```diff
@@ -0,0 +1,69 @@
+db:
+#  build: ../build/mysql
+  image: edxops/mysql:v1
+  environment:
+    - MYSQL_ROOT_PASSWORD=''
+
+mongo:
+  image: edxops/mongo:v2
+  volumes: 
+    - /var/devstack/mongo/data:/data
```
Comment (mrgnr) [inline docker/compose/docker-compose.yml]: We might want to specify volume paths with [variables](https://docs.docker.com/compose/compose-file/#variable-substitution) for more flexibility.

Comment (cpennington) [inline docker/compose/docker-compose.yml]: I think that's not going to be in this PR going forward.

File: docker/build/course_discovery/Dockerfile [hunk 15]
```diff
@@ -0,0 +1,32 @@
+# To build this Dockerfile:
+#
+# From the root of configuration:
+#
+# docker build -f docker/build/course-discovery/Dockerfile .
```
Comment (clintonb) [inline docker/build/course_discovery/Dockerfile]: course_discovery

**Comment by clintonb** [comment]
Note: Getting reacquainted with Docker. We should work to get this on Docker Hub and add documentation for public consumption, once it's ready.

**Comment by clintonb** [comment]
Docs? How do I run this?

File: playbooks/roles/course_discovery/defaults/main.yml (hunk 8 above)
Comment (cpennington) [inline playbooks/roles/course_discovery/defaults/main.yml]: Done.

File: docker/build/course_discovery/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+ARG COURSE_DISCOVERY_VERSION=master
+ARG REPO_OWNER=edx
+
+ADD . /edx/app/edx_ansible/edx_ansible
+
+USER docker
+WORKDIR /edx/app/edx_ansible/edx_ansible/docker/plays
+
+RUN sudo ansible-playbook course_discovery.yml -c local \
+    -t 'install:base,install:code,install:system-requirements,install:app-requirements,install:configuration,install:vhosts,install:devstack' \
```
Comment (e0d) [inline docker/build/course_discovery/Dockerfile]: @cpennington I've started using this ansible_overrides.yml pattern in the build files that I think works pretty well.

https://github.com/edx/configuration/tree/e0d/docker-artifacts/docker/build/forums

The Dockerfile then uses the following:

```
COPY ansible_overrides.yml /
RUN sudo ansible-playbook forum.yml -i '127.0.0.1,' -c local -t "install:base,install:configuration,install:code" -e@/ansible_overrides.yml
```

Comment (cpennington) [inline docker/build/course_discovery/Dockerfile]: I actually don't like that, because it means that the results vary without busting the cache. If the overrides are coded in, either as ARGs or explicit values, then changing them will cause the docker file to rebuild. That's not the case for an external file.

Comment (e0d) [inline docker/build/course_discovery/Dockerfile]: Not true, changing the file will cause the copy to be re-run and the ansible command.

Comment (cpennington) [inline docker/build/course_discovery/Dockerfile]: Oh, I see. Yes, ok, that works then.

**Comment by cpennington** [comment]
@e0d, @clintonb, @feanil: All comments addressed.

**Comment by cpennington** [comment]
@e0d @feanil: Can I get a final review, now that all of the prior commits have been merged?

File: playbooks/roles/course_discovery/defaults/main.yml (hunk 8 above)
Comment (feanil) [inline playbooks/roles/course_discovery/defaults/main.yml]: Is there an update missing here? or are you going to do that separately?

Comment (cpennington) [inline playbooks/roles/course_discovery/defaults/main.yml]: Oh, oops. Thought I had fixed that.

**Comment by feanil** [comment]
I had one more comment on the `course_discovery_requirements` but :+1: either way.
